contracts/TokenProxy.sol

```diff
@@ -0,0 +1,131 @@
+pragma solidity ^0.4.19;
+
+contract TransferableToken {
+    function balanceOf(address _addr) public constant returns (uint256);
+    function transfer(address _to, uint256 _value) public returns (bool success);
+}
+
+
+/// The Gate is a contract with unique address to allow a token holder
+/// (called "User") to transfer tokens from original Token to the Proxy.
+///
+/// The Gate does not know who its User is. The User-Gate relationship is
+/// managed by the Proxy.
+///
+/// TODO: Rename to MigrationGate?
+contract Gate {
+    TransferableToken private TOKEN;
+    address private PROXY;
+
+    /// Gates are to be created by the TokenProxy.
+    function Gate(TransferableToken _token, address _proxy) public {
+        TOKEN = _token;
+        PROXY = _proxy;
+    }
+
+    /// Transfer requested amount of tokens from Gate to Proxy address.
+    /// Only the Proxy can request this and should request transfer of all
+    /// tokens.
+    function transferToProxy(uint256 _value) public {
+        require(msg.sender == PROXY);
+        
+        require(TOKEN.transfer(PROXY, _value));
+    }
+}
+
+
+/// The Proxy for existing tokens implementing a subset of ERC20 interface.
+///
+/// This contract creates a token Proxy contract to extend the original Token
+/// contract interface. The Proxy requires only transfer() and balanceOf()
+/// methods from ERC20 to be implemented in the original Token contract.
+///
+/// All migrated tokens are in Proxy's account on the Token side and distributed
+/// among Users on the Proxy side.
+///
+/// For an user to migrate some amount of ones tokens from Token to Proxy
+/// the procedure is as follows.
+///
+/// 1. Create an individual Gate for migration. The Gate address will be
+///    reported with the GateOpened event and accessible by getGateAddress().
+/// 2. Transfer tokens to be migrated to the Gate address.
+/// 3. Execute Gate.transferToProxy() to finalize the migration.
+///
+/// In the step 3 the User's tokens are going to be moved from the Gate to
+/// the User's balance in the Proxy.
+contract TokenProxy {
+
+    TransferableToken public TOKEN;
+
+    uint256 public totalSupply;
+    mapping(address => uint256) private balances;
+
+    mapping(address => address) private gates;
+
+    event GateOpened(address indexed gate, address indexed user);
+    event GateClosed(address indexed gate, address indexed user);
+
+    // Events taken from ERC777:
+    // TODO: Do you want to keep them like this?
```

Comment:
I think it's not worth having some partial compliance with ERC777 and simple events `Minted/Burn(address, amount)` are enough